### PR TITLE
Add unittest for PixelCNN example.

### PR DESCRIPTION
**Pull request by mohitreddy1996**
* Being done as part of #176 and #231 
* Configure HPs using `ml_collections.ConfigDict`.

### CODE REVIEW

File: examples/pixelcnn/configs/default.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  config = ml_collections.ConfigDict()
+
+  config.learning_rate = 0.001
+  config.lr_decay = 0.999995
+  config.init_batch_size = 16
+  config.batch_size = 64
+  config.num_epochs = 200
+  config.dropout_rate = 0.5
+
+  config.rng = 0
```
Comment (andsteing) [inline examples/pixelcnn/configs/default.py:31]: Maybe rename it to `random_seed` ?

File: examples/pixelcnn/configs/default.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  config.n_resnet = 5
+  config.n_feature = 160
+  config.n_logistic_mix = 10
+
+  config.polyak_decay = 0.9995
+
+  config.num_train_steps = -1
```
Comment (andsteing) [inline examples/pixelcnn/configs/default.py:39]: Maybe add a comment how this relates to `num_epochs`

(See also comment below in `train.py`)

File: examples/pixelcnn/train.py [hunk 3]
```diff
@@ -144,7 +109,7 @@ def loss_fn(model):
   optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
 
   # Compute exponential moving average (aka Polyak decay)
-  ema_decay = FLAGS.polyak_decay
+  ema_decay = polyak_decay
```
Comment (andsteing) [inline examples/pixelcnn/train.py:112]: Consider using `polyak_decay` directly, or renaming the config attribute to `ema_decay`.

File: examples/pixelcnn/train.py [hunk 4]
```diff
@@ -168,40 +133,51 @@ def _prepare(x):
...
+  Args:
+    config: Hyperparameter configuration for training and evaluation.
+    model_dir: Directory to store model data.
+  """
   if jax.host_count() > 1:
     raise ValueError('PixelCNN++ example should not be run on more than 1 host'
                      ' (for now)')
 
+  # Make sure tf does not allocate gpu memory.
+  tf.config.experimental.set_visible_devices([], 'GPU')
```
Comment (andsteing) [inline examples/pixelcnn/train.py:160]: Could we instead move this to the `main()` and `test.setUp()`?
See #472

File: examples/pixelcnn/train.py [hunk 5]
```diff
@@ -210,32 +186,46 @@ def train(pcnn_module, model_dir, batch_size, init_batch_size, num_epochs,
...
-  steps_per_epoch = data_source.TRAIN_IMAGES // batch_size
-  steps_per_eval = data_source.EVAL_IMAGES // batch_size
+  if config.num_train_steps == -1:
+    steps_per_epoch = (
+      data_source.info.splits['train'].num_examples // config.batch_size
+    )
+  else:
+    steps_per_epoch = config.num_train_steps
+
+  if config.num_eval_steps:
```
Comment (andsteing) [inline examples/pixelcnn/train.py:196]: should it be `if config.num_eval_steps == -1:` ?

File: examples/pixelcnn/train.py [hunk 6]
```diff
@@ -210,32 +186,46 @@ def train(pcnn_module, model_dir, batch_size, init_batch_size, num_epochs,
...
 
   # Compute steps per epoch and nb of eval steps
-  steps_per_epoch = data_source.TRAIN_IMAGES // batch_size
-  steps_per_eval = data_source.EVAL_IMAGES // batch_size
+  if config.num_train_steps == -1:
+    steps_per_epoch = (
+      data_source.info.splits['train'].num_examples // config.batch_size
+    )
+  else:
+    steps_per_epoch = config.num_train_steps
```
Comment (andsteing) [inline examples/pixelcnn/train.py:194]: So it is not possible to have `num_train_steps > num_train_examples` ?
I would have assumed that you can specify e.g. `num_train_steps = 2 * num_train_examples` which would result in the same as `num_epochs = 2`.

(I think that's also what I have seen in similar configurations in other projects, though AFAIK we don't use this in Flax examples yet.)

File: examples/pixelcnn/train_test.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  """Test cases for train."""
+
+  def test_train_and_evaluate(self):
+    """Tests pixelcnn training and evaluation loop."""
+    # Create a temporary directory where tensorboard metrics are written.
+    model_dir = tempfile.mkdtemp()
+
+    # Go two directories up to the root of the flax directory.
+    flax_root_dir = pathlib.Path(__file__).parents[2]
+    data_dir = str(flax_root_dir) + '/.tfds/metadata'
```
Comment (andsteing) [inline examples/pixelcnn/train_test.py:39]: Wouldn't it be better to read this from local path `'./.tfds/metadata'` ?

In case tests are run in Github action etc : it's the same, since they're run from the Flax root directory.

In case examples are downloaded independently / forked : There wouldn't be a `../../.tfds/metadata` directory; the test should rather complain with an error message how to get these files - not sure if this is an important usecase, though.

File: pytest.ini [hunk 8]
```diff
@@ -3,7 +3,6 @@
 # UserWarning is due to statement: tensorflow.compat.v2.io import gfile
 # DeprecationWarning is due to statement: tensorflow.compat.v2.io import gfile
 # inspect.getargspec() is invoked in tensorboard/backend/event_processing/event_file_loader.py:61
-# jnp.sum([]) in examples/sst2/train.py:L152 should be replaced with an ndarray.
```
Comment (andsteing) [inline pytest.ini:6]: Is this change related?
(It's fine to keep it, just wanted to make sure it's on purpose)

File: pytest.ini [hunk 9]
```diff
@@ -13,6 +12,5 @@ filterwarnings =
     ignore:can't resolve package from.*:ImportWarning
     ignore:the imp module is deprecated.*:DeprecationWarning
     ignore:inspect.getargspec() is deprecated since Python 3.0.*:DeprecationWarning
-    ignore:jax.numpy reductions won't accept lists and tuples in future versions, only scalars and ndarrays.*:FutureWarning
```
Comment (andsteing) [inline pytest.ini:16]: ditto.

File: pytest.ini [hunk 10]
```diff
@@ -13,6 +12,5 @@ filterwarnings =
...
     ignore:tostring() is deprecated. Use tobytes() instead.*:DeprecationWarning
-    ignore:Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated.*:DeprecationWarning
```
Comment (andsteing) [inline pytest.ini:18]: ditto.

**Review by andsteing — CHANGES_REQUESTED** [review]

File: examples/pixelcnn/configs/default.py (hunk 1 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/configs/default.py]: Done.

File: examples/pixelcnn/configs/default.py (hunk 2 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/configs/default.py:41]: Done, PTAL.

File: examples/pixelcnn/train.py (hunk 3 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/train.py]: Done.

File: examples/pixelcnn/train.py (hunk 4 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/train.py]: Done made the changes, thanks for pointing out. 

Just curious, why can't we have this in `train_and_evaluate`? Since the unit tests, benchmark and even the main would just call that method.

File: examples/pixelcnn/train.py (hunk 5 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/train.py]: Done, thanks!

File: pytest.ini (hunk 8 above)
Comment (mohitreddy1996) [reply pytest.ini:6]: Sorry I think this got added back in one of the PRs which had merge conflicts and I accidentally did a bad merge.

File: examples/pixelcnn/train_test.py (hunk 7 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/train_test.py:45]: So the idea behind this having it this was to have:
- TFDS metadata is downloaded exclusively for these unit tests (we download it automatically in `run_all_tests.sh`) and we place it under flax/.tfds/.. If somebody runs a test individually, they would need to run `download_tfds_metadata` script. Since the downloaded content is specific to Flax, I thought may be it would be best to have in the same directory locality i.e somewhere under flax.

I think we can move this discussion to another forum may be as this will need to be changed (if we agree on something) for other unit tests as well.

File: examples/pixelcnn/train.py (hunk 6 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/train.py:190]: Hmm TBH I actually didn't think of this scenario at all.

Some thoughts:

by `num_train_examples` do you mean `data_source.info.splits['train'].num_examples`? So even with mock data and explicitly specifying `num_examples`, `data_source.info.splits['train'].num_examples` would still return the number of examples in the split information from original metadata i.e 50000 for train split. See warning here: https://www.tensorflow.org/datasets/api_docs/python/tfds/testing/mock_data. 

Also since we do `ds.repeat()` in the input_pipeline, so we can't really do `len(list(ds))`. 

So IMO this shouldn't really effect the training loop?

File: examples/pixelcnn/train.py (hunk 4 above)
Comment (andsteing) [reply examples/pixelcnn/train.py]: We could also have it here, but I think it's cleaner to have this kind of "setup code" in the calling functions.
Don't benchmarks usually call the `main()` function?

File: examples/pixelcnn/train.py (hunk 6 above)
Comment (andsteing) [reply examples/pixelcnn/train.py:190]: Yes, 'data_source.info.splits['train'].num_examples == num_train_examples` (please forgive my laziness :-)

IMO it would be great to have `config.num_train_steps` for total number of training step. This makes it easy to set a target for total number of examples. Alternatively, if `config.num_train_steps == -1`, it would use `num_train_steps = config.num_epochs * steps_per_epoch`. This would allow flexibly to define either in the config, and would also make it possible to stop after a single step in the config regardless of the value of `data_source.info.splits['train'].num_examples`.

(Otherwise, if you prefer keeping the current implementation, I would suggest to rename to `config.num_steps_per_epoch` - also, if it's only used for testing, then you might even use a single config param for this.)

File: examples/pixelcnn/train_test.py (hunk 7 above)
Comment (andsteing) [reply examples/pixelcnn/train_test.py:45]: SGTM, let's move this to a separate discussion.

File: examples/pixelcnn/train.py (hunk 6 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/train.py]: Cool, thanks for the suggestion! Made the necessary changes.

**Review by andsteing — APPROVED** [review]

File: examples/pixelcnn/README.md [hunk 11]
```diff
@@ -24,10 +24,26 @@ The model should run with other configurations and hardware, but was tested on t
...
+```shell
+python train.py \
+--model_dir=./pixelcnn --config=configs/default.py \
+--config.batch_size=320
 ```
-python train.py --batch_size=320
+
+#### 8 x Nvidia V100 (16GB)
+```shell
+python train.py --config=configs/default.py --config.batch_size=320
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply examples/pixelcnn/README.md]: Done, thanks for catching this!

File: examples/pixelcnn/README.md [hunk 12]
```diff
@@ -24,10 +24,26 @@ The model should run with other configurations and hardware, but was tested on t
...
 
 ### How to run
-#### 8 x Nvidia V100 (16GB)
-To run training
+
+`python train.py --model_dir=./pixelcnn`
+
+#### Overriding Hyperparameter configurations
+
+MNIST example allows specifying a hyperparameter configuration by the means of
```
Comment (avital) [inline examples/pixelcnn/README.md:32]: This isn't MNIST

File: examples/pixelcnn/README.md [hunk 13]
```diff
@@ -24,10 +24,26 @@ The model should run with other configurations and hardware, but was tested on t
...
+setting `--config` flag. Configuration flag is defined using
+[config_flags](https://github.com/google/ml_collections/tree/master#config-flags).
+`config_flags` allows overriding configuration fields. This can be done as
+follows:
+
+```shell
+python train.py \
+--model_dir=./pixelcnn --config=configs/default.py \
```
Comment (avital) [inline examples/pixelcnn/README.md:40]: Is `--config` mandatory?

File: examples/pixelcnn/README.md [hunk 14]
```diff
@@ -24,10 +24,26 @@ The model should run with other configurations and hardware, but was tested on t
...
+--config.num_epochs=100
 ```
-python train.py --batch_size=320
+
+#### 8 x Nvidia V100 (16GB)
+```shell
+python train.py --model_dir=./pixelcnn --config=configs/default.py --config.batch_size=320
```
Comment (avital) [inline examples/pixelcnn/README.md:46]: I think we should not have `configs/default.py`. We should only have `configs/v100_x8.py` (then we can add `configs/v100.py`, ...

File: examples/pixelcnn/configs/testing.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Testing Hyperparameter configuration."""
+
+from configs import default as default_lib
+
+
+def get_config():
```
Comment (avital) [inline examples/pixelcnn/configs/testing.py:20]: I like that we have a separate file for a testing config.

File: examples/pixelcnn/README.md (hunk 14 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/README.md]: Great suggestion, done!

File: examples/pixelcnn/README.md (hunk 13 above)
Comment (mohitreddy1996) [reply examples/pixelcnn/README.md]: Done.

File: examples/pixelcnn/configs/default.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def get_config():
+  """Get the default hyperparameter configuration."""
+  config = ml_collections.ConfigDict()
+
+  config.learning_rate = 0.001
+  config.lr_decay = 0.999995
+  config.init_batch_size = 16
+  config.batch_size = 64
```
Comment (avital) [inline examples/pixelcnn/configs/default.py:27]: cc @levskaya what do you think about configuration with defaults that are overriden like this, vs having less (or nothing) in a default config and specifying more explicitly in each separate config file?

**Comment by avital** [comment]
This PR is on the old pre-Linen example, so I'll close it.

I see that linen_examples/pixelcnn has a test: https://github.com/google/flax/blob/master/linen_examples/pixelcnn/train_test.py#L43, is that sufficient?
